src/opencl/mod.rs

```diff
@@ -201,121 +173,174 @@ fn get_device_by_index(index: usize) -> Option<&'static Device> {
+
+    pub fn from_opencl(device: &Device, src: &str) -> GPUResult<Program> {
+        let cached = utils::cache_path(device, src)?;
         if std::path::Path::exists(&cached) {
             let bin = std::fs::read(cached)?;
             Program::from_binary(device, bin)
         } else {
-            let context = ocl::Context::builder()
-                .platform(device.platform)
-                .devices(device.device)
-                .build()?;
-            let program = ocl::Program::builder()
-                .src(src)
-                .devices(ocl::builders::DeviceSpecifier::Single(device.device))
-                .build(&context)?;
-            let queue = ocl::Queue::new(&context, device.device, None)?;
+            let context = Context::from_device(&device.device)?;
+            let mut program = opencl3::program::Program::create_from_source(&context, src)?;
+            if let Err(build_error) = program.build(context.devices(), "") {
+                let log = program.get_build_log(context.devices()[0])?;
+                return Err(GPUError::Opencl3(build_error, Some(log)));
+            }
+            let queue = CommandQueue::create(&context, context.default_device(), 0)?;
+            let kernels = opencl3::kernel::create_program_kernels(&program)?;
+            let kernels_by_name = kernels
+                .into_iter()
+                .map(|kernel| {
+                    let name = kernel.function_name()?;
+                    Ok((name, kernel))
+                })
+                .collect::<Result<_, ClError>>()?;
             let prog = Program {
-                program,
+                device_name: device.name(),
                 queue,
-                device,
+                context,
+                kernels: kernels_by_name,
             };
-            std::fs::write(cached, prog.to_binary()?)?;
+            let binaries = program
+                .get_binaries()
+                .map_err(|_| GPUError::ProgramInfoNotAvailable(CL_PROGRAM_BINARIES))?;
+            std::fs::write(cached, binaries[0].clone())?;
             Ok(prog)
         }
     }
-    pub fn from_binary(device: Device, bin: Vec<u8>) -> GPUResult<Program> {
-        let context = ocl::Context::builder()
-            .platform(device.platform)
-            .devices(device.device)
-            .build()?;
+
+    pub fn from_binary(device: &Device, bin: Vec<u8>) -> GPUResult<Program> {
+        let context = Context::from_device(&device.device)?;
         let bins = vec![&bin[..]];
-        let program = ocl::Program::builder()
-            .binaries(&bins)
-            .devices(ocl::builders::DeviceSpecifier::Single(device.device))
-            .build(&context)?;
-        let queue = ocl::Queue::new(&context, device.device, None)?;
+        let mut program =
+            opencl3::program::Program::create_from_binary(&context, context.devices(), &bins)?;
+        if let Err(build_error) = program.build(context.devices(), "") {
```

Comment:
this part seems to be the same as in `from_opencl`, could this code be shared?

Reply:
Also large parts below are the same, but refactoring it didn't work out.

Reply:
ok